chain/websocketd.py

```diff
@@ -23,29 +23,96 @@
 
 logger.info("websocketd.py started")
 
+zmq_socks = set()
+zmq_socks_tag = {}
+tag_zmq_sock = {}
+tag_subscribers = {}
 
 @websockets.route('/<tag>')
 def site_socket(ws, tag):
     logger.info('ws client connected for tag "%s"' % tag)
-    zmq_sock = zmq_ctx.socket(zmq.SUB)
-    zmq_sock.connect(ZMQ_PASSTHROUGH_URL_PUB)
-    # note that flask gives us tag as a unicode string
-    zmq_sock.setsockopt_string(zmq.SUBSCRIBE, tag)
+    if tag in tag_zmq_sock:
+        tag_subscribers[tag].add(ws)
+    else:
+        zmq_sock = zmq_ctx.socket(zmq.SUB)
+        zmq_sock.connect(ZMQ_PASSTHROUGH_URL_PUB)
+        # note that flask gives us tag as a unicode string
+        zmq_sock.setsockopt_string(zmq.SUBSCRIBE, tag)
+        zmq_socks.add(zmq_sock)
+        zmq_socks_tag[zmq_sock] = tag
+        tag_zmq_sock[tag] = zmq_sock
+        tag_subscribers[tag] = {ws}
     while True:
-        in_data = zmq_sock.recv()
-        # in case we later subscribe to multiple topic, get the topic from the
-        # incoming message
-        msg_tag, _, msg = in_data.partition(' ')
-        logger.debug('Received on tag "%s": %s' % (msg_tag, msg))
         try:
-            ws.send(msg)
+            # Read from ws socket continuously.
+            # XXX:  This builds up a lot of threads.  Is there ANY way we could do
```

Comment:
Does this actually spawn system threads, or do you mean greenlets? As I understand it flask-sockets is based on gevent-websocket, which should spawn a greenlet per connection (which will be "green-blocked" on `ws.receive()`) Under the hood I'd expect it do be doing something like `select`